posts/2022-03-01-user-feature-install.adoc

```diff
@@ -0,0 +1,378 @@
+
+.pom.xml
+[source, xml]
+----
+<project>
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>test.user.test.osgi</groupId>
+  <artifactId>SimpleActivatorESA-</artifactId>
+  <version>1.0.0-SNAPSHOT</version>
+
+  <packaging>esa</packaging> <!-- set packaging type to esa -->
+
+  <dependencies>
+    <!-- Add OSGi bundle -->
+    <dependency>
+		<groupId>test.user.test.osgi</groupId>
+    	<artifactId>SimpleActivator</artifactId>
+    	<version>0.0.1-SNAPSHOT</version>
+	</dependency>
+  </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.aries</groupId>
+        <artifactId>esa-maven-plugin</artifactId>
+        <version>1.0.0</version>
+        <extensions>true</extensions>
+        <configuration>
+          <generateManifest>true</generateManifest>
+          <archiveContent>all</archiveContent>
+          <instructions>
+            <Subsystem-Vendor>IBM</Subsystem-Vendor>
+            <IBM-Feature-Version>2</IBM-Feature-Version>
+            <IBM-ShortName>SimpleActivator</IBM-ShortName>
+            <Subsystem-Type>osgi.subsystem.feature</Subsystem-Type>
+            <Subsystem-SymbolicName>
+                com.example.bundle.Activator;visibility:=public
+            </Subsystem-SymbolicName>
+            <Subsystem-Version>1.0.0</Subsystem-Version>
+          </instructions>
+        </configuration>
+      </plugin>
+    </plugins>
+  </build>
+</project>
+----
+By default it will not generate a manifest, so we have to set the `generateManifest` header to *true*. Note that in order to install the feature, we need to set the *visibility* directive to "public". We can do so by setting the `Subsystem-SymbolicName` header to "_Bundle-SymbolicName;visibility:=public_". If the *visibility* is set to `protected|private`, Liberty Maven Plugin will not be able to resolve the feature. Also, the mandatory header `Subsystem-Content` will be created automatically by the plugin. 
+
+
+
+Run `**_mvn clean install_**` to create an ESA file. Inside the ESA file, you will find your bundle JAR and SUBSYSTEM.MF.
+
+.SUBSYSTEM.MF
+[source, txt]
+----
+Subsystem-ManifestVersion: 1
+Subsystem-SymbolicName: com.example.bundle.Activator;visibility:=public
+Subsystem-Version: 1.0.0
+Subsystem-Name: SimpleActivatorESA
+Subsystem-Content: com.example.bundle.Activator;version="[1.0.0,1.0.0]"
+IBM-Feature-Version: 2
+IBM-ShortName: SimpleActivator
+Subsystem-Type: osgi.subsystem.feature
+Subsystem-Vendor: IBM
+----
+'''
+=== Creating Bill of Materials (BOM) 
+Create a Bill of Materials (BOM) for the user feature ESA file. The BOM is a pom file that manages the dependencies of the project. Liberty Maven Plugin's `prepare-feature` goal and `install-feature` goal require the BOM file to install the Liberty user feature onto Liberty runtime. 
```

Comment:
Liberty Maven Plugin's `prepare-feature` goal and `install-feature` goal require the BOM file to install the Liberty user feature onto Liberty runtime.

--->

The Liberty Maven Plugin `prepare-feature` and `install-feature` goals require a BOM file to install a Liberty user feature onto the Liberty runtime.